tf2onnx/onnx_opset/nn.py

```diff
@@ -623,6 +623,20 @@ def _convert(cls, ctx, node, **kwargs):
         else:
             spatial = 2
 
+        origin_dtype = ctx.get_dtype(node.output[0])
+        if origin_dtype not in [onnx_pb.TensorProto.FLOAT16, onnx_pb.TensorProto.FLOAT,
+                                onnx_pb.TensorProto.DOUBLE]:
+            # the onnx spec doesn't allow int types for pool ops
+            cast_node = ctx.insert_new_node_on_input(node, "Cast", node.input[0], to=onnx_pb.TensorProto.FLOAT)
+            ctx.set_dtype(cast_node.output[0], onnx_pb.TensorProto.FLOAT)
+            ctx.copy_shape(node.name, cast_node.output[0])
+
+            cast_back_node = ctx.insert_new_node_on_output("Cast", node.output[0],
+                                                           name=utils.make_name(node.name) + "_castback",
+                                                           to=origin_dtype)
+            ctx.set_dtype(cast_back_node.output[0], origin_dtype)
+            ctx.copy_shape(node.name, cast_back_node.output[0])
```

Comment:
I've been trying to avoid using set_dtype and copy_shape for new nodes since it triggers shape inference and then overwrites the result.  You can use make_node (with shapes and dtypes set) and insert_node_on_output, or add a shapes/dtypes arg to insert_new_node_on_output